### PR TITLE
feat(gatsby): make dev ssr bundling lazy

### PR DESCRIPTION
**Pull request by KyleAMathews**
Similar to lazy client bundling PR https://github.com/gatsbyjs/gatsby/pull/27884/

Saves a significant amount of time on bootup

### CODE REVIEW

File: packages/gatsby/src/utils/dev-ssr/render-dev-html.ts [hunk 1]
```diff
@@ -63,6 +128,25 @@ export const renderDevHTML = ({
...
+
+    // We'll also get a head start on compiling the client code (this
+    // call has no effect if the page component is already in the client bundle).
+    createClientVisitedPage(pageObj.componentChunkName)
+
+    // Ensure the query has been run and written out.
+    await getPageDataExperimental(pageObj.path)
+
+    // Wait for public/render-page.js to update w/ the page component.
+    await ensurePathComponentInSSRBundle(pageObj, directory)
```
Comment (KyleAMathews) [inline packages/gatsby/src/utils/dev-ssr/render-dev-html.ts:148]: this function is the meat of the change

File: packages/gatsby/src/bootstrap/requires-writer.ts [hunk 2]
```diff
@@ -184,11 +186,27 @@ export const writeAll = async (state: IGatsbyState): Promise<boolean> => {
   const pages = [...state.pages.values()]
   const matchPaths = getMatchPaths(pages)
   const components = getComponents(pages)
+  let cleanedSSRVisitedPageComponents: Array<IGatsbyPageComponent> = components
```
Comment (vladar) [inline packages/gatsby/src/bootstrap/requires-writer.ts:189]: Is there a reason to assign `components` here? We later pass `cleanedSSRVisitedPageComponents` to `createHash` below (line 233) and so this affects the current stable codepath - without the flag we pass all components to `createHash`.

Probably not critical but seems unnecessary.

Comment (KyleAMathews) [reply packages/gatsby/src/bootstrap/requires-writer.ts:189]: This was the simplest way I could see to make typescript happy 🤷 What's the better way to assign a type here?

Comment (vladar) [reply packages/gatsby/src/bootstrap/requires-writer.ts:189]: I think empty array should work?

**Review by vladar — APPROVED** [review]

**Pull request merged**